Minimize Vimscript

Iron-E opened the pull request:
This PR aims to minimize the amount of Vimscript contained in the project. Old publicly-facing code (e.g. `bufferline#enable()`) is kept to ensure that users do not have to change their configurations.

Closes #178

Iron-E commented on autoload/bufferline.vim:
```diff
@@ -0,0 +1,51 @@
+function! bufferline#enable()
```
All of these functions are still there, but just point to the Lua `bufferline` module now. If I got rid of them it might break something someone's made :grimacing:

Iron-E commented on lua/bbye.lua:
```diff
@@ -0,0 +1,180 @@
+--- Use `vim.notify` to print an error `msg`
+--- @param msg string
+local function err(msg)
+  vim.notify(msg, vim.log.levels.ERROR, {title = 'bbye'})
```
`title` is not an official part of `vim.notify`, but that will give it integration with the `nvim-notify` plugin.

Iron-E commented on lua/bufferline/render.lua:
```diff
@@ -307,13 +307,13 @@ local function render(update_names)
       if click_enabled then
         closePrefix =
-            '%' .. buffer_number .. '@BufferlineCloseClickHandler@' .. closePrefix
+            '%' .. buffer_number .. '@v:lua.BufferlineCloseClickHandler@' .. closePrefix
```
I haven't tested this but it should work.

Iron-E commented on lua/bufferline/highlight.lua:
```diff
@@ -0,0 +1,95 @@
+--- @return string color
+local function color(index, groups, default)
+  for _, group in ipairs(groups) do
+    local hl = vim.api.nvim_get_hl_by_name(group, true)
```
I can probably sneak in #207 here by passing `vim.go.termguicolors` instead of `true`.

romgrk commented on README.md:
```diff
@@ -504,5 +504,5 @@ No, barbar has nothing to do with barbarians.
-barbar.nvim: Distributed under the terms of the JSON license.  
-bbye.vim: Distributed under the terms of the GNU Affero license.  
+barbar.nvim: Distributed under the terms of the JSON license.
+bbye.vim: Distributed under the terms of the GNU Affero license.
```
Trailing spaces are required to split the two lines.

romgrk commented on autoload/bufferline.vim:
```diff
@@ -0,0 +1,51 @@
+
+function! bufferline#close_direct(abuf)
+   call luaeval("require'bufferline.state'.close_buffer(_A)", a:abuf)
+endfunc
```
For the indentation, not sure if it's github but I'm seeint mixed tabs & 3-spaces indentation. Can we keep all on 2-spaces?

romgrk commented on autoload/bufferline/bbye.vim:
```diff
@@ -24,111 +24,5 @@
-    echomsg a:msg
-    echohl NONE
-    let v:errmsg = a:msg
+  lua require'bbye'.delete(a:action, a:bang, a:buffer_name, ...)
```
Maybe keep the `bufferline.bbye` namespace? Just to be extra sure we don't conflict with anything every.

romgrk commented on lua/bufferline/highlight.lua:
```diff
@@ -0,0 +1,95 @@
+  --       -Index: buffer index
+  --         -Mod: when modified
+  --        -Sign: the separator between buffers
+  --      -Target: letter in buffer-picking mode
```
Here maybe add `local set_hl = vim.api.nvim_set_hl`? Easier to read, less noise.

romgrk commented on lua/bufferline/state.lua:
```diff
@@ -574,19 +574,19 @@ end
+  for _, number in ipairs(buffers) do
     if number ~= current then
-      vim.fn['bufferline#bbye#delete']('bdelete', '', bufname(number))
+      require'bbye'.delete('bdelete', false, bufname(number))
```
As we're now using lua here, we should add `local delete = requre'...'` at the start of the file, with the other imports.

romgrk commented:
LGTM overall, once you've finished with the comments you & I left, we should go through the options and test that they're all still working correctly. It's kinda tedious to check them all but it avoids having broken stuff on user's setups.

Iron-E commented:
> Edit: I've pulled the PR, here is what I see. Lmk if you can't reproduce.

Are you referring to the `[no name]` which is highlighted differently than the others? I couldn't reproduce that on my side:

![cap](https://user-images.githubusercontent.com/36409591/167890976-4690f5d1-238a-4665-b91a-fbdcb67788e6.png)

If you can get it to happen again, could you tell me what `&tabline` was?

romgrk commented on plugin/bufferline.lua:
```diff
@@ -0,0 +1,160 @@
+
+--------------------
+-- Section: Commands
+--------------------
```
For this section, can we extract `local create_command = vim.api.nvim_create_user_command`? Very verbose otherwise

romgrk commented:
The previous issue has disappeared with the latest version of this PR. I'll use it over the weekend and report if there are other issues. Do you have more stuff you want to do for this PR?

Iron-E commented:
I added back the requested `local foo = vim.api.nvim_xyz`, but also moved the `plugin/bufferline.lua` code into `bufferline.setup`. My thinking is that we get two benefits:

1. If we ever wanted to do lazy initialization like other Lua plugins are doing these days, this puts us closer to that.
2. The `local foo = vim.api.nvim_xyz` are function-local and so will not hang around in memory after being set (best of both worlds!).

The change is purely on our side— the plugin still initializes automatically.

melkster reviewed (approved):
Great work!

Iron-E commented:
I was able to reproduce that bug (bright red highlighting for effect):

![cap](https://user-images.githubusercontent.com/36409591/168696858-8bcff3f3-4a1e-402e-b56a-e2115359f422.png)

`&tabline` is:

```log
%#BufferCurrentSign#▊ %#DevIconRsCurrent# %#BufferCurrent#employee_adapter.rs %#BufferCurrent# %#BufferInactiveSign#▎%#BufferInactive# %#BufferInactive#ex
pense_adapter.rs %#BufferInactive# %#BufferInactiveSign#▎%#BufferInactive# %#BufferInactive#mod.rs %#BufferInactive# %#BufferInactiveSign#▎%#BufferInactiv
e# %#BufferInactive#contact_info/contact_info_adapter.rs %#BufferInactive# %#BufferVisibleSign#▊ %#DevIconRsVisible# %#BufferVisible#expense.rs %#BufferVi
sible# %#BufferInactiveSign#▎%#BufferInactive# %#BufferInactive#timesheet.rs %#BufferInactive# %#BufferInactiveSign#▎%#BufferInactive# %#BufferInactive#L
ocalPostg%0@BufferlineMainClickHandler@%#BufferTabpageFill#%=%#BufferTabpages# 1/2 %#BufferTabpageFill#
```

Iron-E commented:
I reproduced the bug on `master`, so I think it is safe to merge this PR. I will try to fix it as part of #229.

![cap](https://user-images.githubusercontent.com/36409591/168697063-93e69b7d-242d-44f3-800c-0870a77017d1.png)

Iron-E commented:
I'll merge this for now and we can follow up in #229 as necessary
